### PR TITLE
Add Span::unique_site

### PR DESCRIPTION
**Pull request by mjbshaw**
A little bit of background: [Feedback on Span::def_site](https://internals.rust-lang.org/t/feedback-on-span-def-site/9831)

`Span::def_site` is useful in that call-site code cannot access identifiers which use it. But it doesn't prevent identifier clashes within the macro, particularly when the macro is implemented using several subcomponents, each of which generates a portion of the macro. These subcomponents may use conflicting identifiers.

The new `Span::unique_site` solves this problem. Each call to `Span::unique_site` generates a new span that resolves at a unique site. It allows a macro to isolate parts of its expansion from each other.

### CODE REVIEW

File: src/libproc_macro/lib.rs [hunk 1]
```diff
@@ -255,6 +255,13 @@ macro_rules! diagnostic_method {
 }
 
 impl Span {
+    /// A span that resolves at a unique site. Each call to this method will return a new unique
+    /// span that is inaccessible from other sites.
```
Comment (Centril) [inline src/libproc_macro/lib.rs]: The documentation should mention "gensym" here in some fashion since that's familiar for some folks who would use this.

Comment (mjbshaw) [reply src/libproc_macro/lib.rs]: Good idea.

**Comment by petrochenkov** [comment]
I need to think how to do this properly, so it may take some time.

Adding a new `Transparency` variant doesn't seem right, and giving the power to create `Symbol::gensym`s could interfere with compiler's implementation details.

**Comment by petrochenkov** [comment]
I'm going to close this, I think.
This needs an RFC including motivation and some more detailed design / specification.

How to approach writing such an RFC.
1) Audit all uses of `gensym` in the compiler and figure out which uses could be satisfied by `Span::def_site` and which could not.
2) Describe use cases that cannot be satisfied by `Span::def_site` in detail.
3) Attempt to address those use cases with `Span::unique_site` introduced as a `Span::def_site` of a "freshly introduced macro" and see whether it fits.
4) Attempt to address use cases from outside the compiler with `Span::unique_site` introduced as a `Span::def_site` of a "freshly introduced macro" and see whether it fits.

The motivation for introducing `Span::unique_site` through `Span::def_site` is to have one mechanism for gensyms-and-similar-things rather than two, and to get rid of the existing gensym implementation in the compiler.

**Comment by mjbshaw** [comment]
@petrochenkov Sorry, I'm not sure I fully understand your comment. I'm not sure if `Span::unique_site` is useful within the compiler, and I have no idea how `gensym` is used within the compiler. My motivation is to use `Span::unique_site` in user code (procedural macros, specifically), much like I'd want to use `gensym` in Lisp (or Racket, which is the Lisp dialect I have experience with). There is no `gensym` mechanism that I'm familiar with that is available from user code for procedural macros.

I can write an RFC, but I'm kinda surprised you want a whole RFC for a single method on a type that is already 90% unstable. But sure, I'll write an RFC.

**Comment by petrochenkov** [comment]
@mjbshaw 
>and I have no idea how gensym is used within the compiler.

For generating unique identifiers, for various purposes.
That's what you want from `Span::unique_span` more or less.
However, for some of those purposes `Span::def_site` should be enough.

>My motivation is to use Span::unique_site in user code (procedural macros, specifically), much like I'd want to use gensym in Lisp

It would be great to describe those use cases in the RFC.

>I'm kinda surprised you want a whole RFC for a single method on a type that is already 90% unstabl

I wouldn't underestimate how tricky and complex this single method is.

**Pull request closed without merging**